File: app/modules/planning/database/planner_repository.py
"""Planner repository for holistic planner sequences and evaluations."""

import json
import logging
from datetime import datetime
from typing import Dict, List, Optional

from app.core.database.manager import get_db_manager
from app.modules.planning.domain.holistic_planner import ActionCandidate

logger = logging.getLogger(__name__)


class PlannerRepository:
    """Repository for planner database operations."""

    def __init__(self):
        self._db_manager = None

    async def _get_db(self):
        """Get planner database connection."""
        if self._db_manager is None:
            self._db_manager = get_db_manager()
        return self._db_manager.planner

    async def ensure_sequences_generated(
        self, portfolio_hash: str, sequences: List[List[ActionCandidate]]
    ) -> None:
        """
        Check if sequences exist for portfolio_hash, if not insert all sequences.

        Args:
            portfolio_hash: Portfolio hash
            sequences: List of sequences to insert (each is List[ActionCandidate])
        """
        db = await self._get_db()

        # Check if sequences already exist
        row = await db.fetchone(
            "SELECT COUNT(*) as count FROM sequences WHERE portfolio_hash = ?",
            (portfolio_hash,),
        )
        if row and row["count"] > 0:
            logger.debug(
                f"Sequences already exist for portfolio {portfolio_hash[:8]}..."
            )
            return

        # Insert all sequences
        now = datetime.now().isoformat()
        inserted = 0

        for sequence in sequences:
            if not sequence:
                continue

            # Calculate sequence hash
            sequence_repr = [(c.symbol, c.side, c.quantity) for c in sequence]
            sequence_json_str = json.dumps(sequence_repr, sort_keys=False)
            import hashlib

            sequence_hash = hashlib.md5(sequence_json_str.encode()).hexdigest()

            # Calculate priority (sum of action priorities)
            priority = sum(c.priority for c in sequence)

            # Determine pattern type (simplified - could be enhanced)
            pattern_type = self._determine_pattern_type(sequence)

            # Serialize sequence
            sequence_data = [
                {
                    "side": c.side,
                    "symbol": c.symbol,
                    "name": c.name,
                    "quantity": c.quantity,
                    "price": c.price,
                    "value_eur": c.value_eur,
                    "currency": c.currency,
                    "priority": c.priority,
                    "reason": c.reason,
                    "tags": c.tags,
                }
                for c in sequence
            ]
            sequence_json = json.dumps(sequence_data)

            await db.execute(
                """INSERT INTO sequences
                   (sequence_hash, portfolio_hash, priority, sequence_json, depth, pattern_type, completed, created_at)
                   VALUES (?, ?, ?, ?, ?, ?, 0, ?)""",
                (
                    sequence_hash,
                    portfolio_hash,
                    priority,
                    sequence_json,
                    len(sequence),
                    pattern_type,
                    now,
                ),
            )
            inserted += 1

        await db.commit()
        logger.info(
            f"Inserted {inserted} sequences for portfolio {portfolio_hash[:8]}..."
        )

    def _determine_pattern_type(self, sequence: List[ActionCandidate]) -> str:
        """Determine pattern type from sequence (simplified heuristic)."""
        if len(sequence) == 1:
            return "single_best"
        has_sell = any(c.side == "SELL" for c in sequence)
        has_buy = any(c.side == "BUY" for c in sequence)
        if has_sell and has_buy:
            return "mixed_strategy"
        elif has_sell:
            return "multi_sell"
        else:
            return "multi_buy"

    async def get_next_sequences(
        self, portfolio_hash: str, limit: int = 100
    ) -> List[Dict]:
        """
        Get next N sequences ordered by priority DESC, not completed.

        Args:
            portfolio_hash: Portfolio hash
            limit: Maximum number of sequences to return

        Returns:
            List of dicts with sequence_hash, sequence_json, priority, depth, pattern_type
        """
        db = await self._get_db()

        rows = await db.fetchall(
            """SELECT sequence_hash, sequence_json, priority, depth, pattern_type
               FROM sequences
               WHERE portfolio_hash = ? AND completed = 0
               ORDER BY priority DESC
               LIMIT ?""",
            (portfolio_hash, limit),
        )

        return [dict(row) for row in rows]

    async def mark_sequence_completed(
        self, sequence_hash: str, portfolio_hash: str, evaluated_at: str
    ) -> None:
        """
        Mark sequence as completed.

        Args:
            sequence_hash: Sequence hash
            portfolio_hash: Portfolio hash
            evaluated_at: ISO timestamp when evaluated
        """
        db = await self._get_db()

        await db.execute(
            """UPDATE sequences
               SET completed = 1, evaluated_at = ?
               WHERE sequence_hash = ? AND portfolio_hash = ?""",
            (evaluated_at, sequence_hash, portfolio_hash),
        )
        await db.commit()

    async def has_evaluation(self, sequence_hash: str, portfolio_hash: str) -> bool:
        """
        Check if evaluation exists for sequence.

        Args:
            sequence_hash: Sequence hash
            portfolio_hash: Portfolio hash

        Returns:
            True if evaluation exists, False otherwise
        """
        db = await self._get_db()

        row = await db.fetchone(
            "SELECT COUNT(*) as count FROM evaluations WHERE sequence_hash = ? AND portfolio_hash = ?",
            (sequence_hash, portfolio_hash),
        )

        return row and row["count"] > 0

    async def insert_evaluation(
        self,
        sequence_hash: str,
        portfolio_hash: str,
        end_score: float,
        breakdown: Dict,
        end_cash: float,
        end_context_positions: Dict[str, float],
        div_score: float,
        total_value: float,
    ) -> None:
        """
        Insert evaluation result.

        If evaluation already exists, skip insertion (avoids PRIMARY KEY violation).

        Args:
            sequence_hash: Sequence hash
            portfolio_hash: Portfolio hash
            end_score: Final portfolio score
            breakdown: Score breakdown dict
            end_cash: Cash remaining after sequence
            end_context_positions: Final positions dict
            div_score: Diversification score
            total_value: Total portfolio value
        """
        # Check if evaluation already exists
        if await self.has_evaluation(sequence_hash, portfolio_hash):
            logger.debug(
                f"Evaluation already exists for sequence {sequence_hash[:8]}, skipping insertion"
            )
            return

        db = await self._get_db()
        now = datetime.now().isoformat()

        await db.execute(
            """INSERT INTO evaluations
               (sequence_hash, portfolio_hash, end_score, breakdown_json, end_cash,
                end_context_positions_json, div_score, total_value, evaluated_at)
               VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)""",
            (
                sequence_hash,
                portfolio_hash,
                end_score,
                json.dumps(breakdown),
                end_cash,
                json.dumps(end_context_positions),
                div_score,
                total_value,
                now,
            ),
        )
        await db.commit()

    async def get_best_result(self, portfolio_hash: str) -> Optional[Dict]:
        """
        Get best result from best_result table.

        Args:
            portfolio_hash: Portfolio hash

        Returns:
            Dict with best_sequence_hash, best_score, updated_at or None
        """
        db = await self._get_db()

        row = await db.fetchone(
            "SELECT best_sequence_hash, best_score, updated_at FROM best_result WHERE portfolio_hash = ?",
            (portfolio_hash,),
        )

        if row:
            return dict(row)
        return None

    async def update_best_result(
        self, portfolio_hash: str, sequence_hash: str, score: float
    ) -> None:
        """
        Insert or update best_result table.

        Args:
            portfolio_hash: Portfolio hash
            sequence_hash: Best sequence hash
            score: Best score
        """
        db = await self._get_db()
        now = datetime.now().isoformat()

        await db.execute(
            """INSERT OR REPLACE INTO best_result
               (portfolio_hash, best_sequence_hash, best_score, updated_at)
               VALUES (?, ?, ?, ?)""",
            (portfolio_hash, sequence_hash, score, now),
        )
        await db.commit()

    async def delete_sequences_only(self, portfolio_hash: str) -> None:
        """
        Delete sequences only (keep evaluations and best_result).

        Used when regenerating sequences with new settings - evaluations
        are still valid and can be reused if sequences overlap.

        Args:
            portfolio_hash: Portfolio hash
        """
        db = await self._get_db()

        await db.execute(
            "DELETE FROM sequences WHERE portfolio_hash = ?", (portfolio_hash,)
        )
        await db.commit()

        logger.info(
            f"Deleted sequences (kept evaluations) for portfolio {portfolio_hash[:8]}..."
        )

    async def delete_sequences_for_portfolio(self, portfolio_hash: str) -> None:
        """
        Delete all sequences, evaluations, and best_result for portfolio_hash.

        Args:
            portfolio_hash: Portfolio hash
        """
        db = await self._get_db()

        await db.execute(
            "DELETE FROM sequences WHERE portfolio_hash = ?", (portfolio_hash,)
        )
        await db.execute(
            "DELETE FROM evaluations WHERE portfolio_hash = ?", (portfolio_hash,)
        )
        await db.execute(
            "DELETE FROM best_result WHERE portfolio_hash = ?", (portfolio_hash,)
        )
        await db.commit()

        logger.info(f"Deleted all planner data for portfolio {portfolio_hash[:8]}...")

    async def has_sequences(self, portfolio_hash: str) -> bool:
        """
        Check if sequences exist for portfolio_hash.

        Args:
            portfolio_hash: Portfolio hash

        Returns:
            True if sequences exist, False otherwise
        """
        db = await self._get_db()

        row = await db.fetchone(
            "SELECT COUNT(*) as count FROM sequences WHERE portfolio_hash = ?",
            (portfolio_hash,),
        )

        return row and row["count"] > 0

    async def get_evaluation_count(self, portfolio_hash: str) -> int:
        """
        Get count of evaluated sequences for portfolio.

        Args:
            portfolio_hash: Portfolio hash

        Returns:
            Number of evaluated sequences
        """
        db = await self._get_db()

        row = await db.fetchone(
            "SELECT COUNT(*) as count FROM evaluations WHERE portfolio_hash = ?",
            (portfolio_hash,),
        )

        return row["count"] if row else 0

    async def get_total_sequence_count(self, portfolio_hash: str) -> int:
        """
        Get total count of sequences for portfolio.

        Args:
            portfolio_hash: Portfolio hash

        Returns:
            Total number of sequences
        """
        db = await self._get_db()

        row = await db.fetchone(
            "SELECT COUNT(*) as count FROM sequences WHERE portfolio_hash = ?",
            (portfolio_hash,),
        )

        return row["count"] if row else 0

    async def are_all_sequences_evaluated(self, portfolio_hash: str) -> bool:
        """
        Check if all sequences for portfolio_hash have been evaluated.

        Args:
            portfolio_hash: Portfolio hash

        Returns:
            True if all sequences are evaluated, False otherwise
        """
        db = await self._get_db()

        row = await db.fetchone(
            """SELECT COUNT(*) as total, SUM(completed) as completed
               FROM sequences WHERE portfolio_hash = ?""",
            (portfolio_hash,),
        )

        if not row or row["total"] == 0:
            return False

        return row["completed"] == row["total"]

    async def get_best_sequence_from_hash(
        self, portfolio_hash: str, sequence_hash: str
    ) -> Optional[List[ActionCandidate]]:
        """
        Get sequence from hash and deserialize to List[ActionCandidate].

        Args:
            portfolio_hash: Portfolio hash
            sequence_hash: Sequence hash

        Returns:
            List[ActionCandidate] or None if not found
        """
        db = await self._get_db()

        row = await db.fetchone(
            "SELECT sequence_json FROM sequences WHERE sequence_hash = ? AND portfolio_hash = ?",
            (sequence_hash, portfolio_hash),
        )

        if not row:
            return None

        sequence_data = json.loads(row["sequence_json"])

        return [
            ActionCandidate(
                side=c["side"],
                symbol=c["symbol"],
                name=c["name"],
                quantity=c["quantity"],
                price=c["price"],
                value_eur=c["value_eur"],
                currency=c["currency"],
                priority=c["priority"],
                reason=c["reason"],
                tags=c["tags"],
            )
            for c in sequence_data
        ]

File: trader/internal/modules/planning/repository/planner_repository.go
package repository

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"time"

	"github.com/aristath/arduino-trader/internal/database"
	"github.com/aristath/arduino-trader/internal/modules/planning/domain"
	"github.com/rs/zerolog"
)

// PlannerRepository handles database operations for planning module.
// Database: agents.db (sequences, evaluations, best_result tables)
type PlannerRepository struct {
	db  *database.DB // agents.db
	log zerolog.Logger
}

// NewPlannerRepository creates a new planner repository.
// db parameter should be agents.db connection
func NewPlannerRepository(db *database.DB, log zerolog.Logger) *PlannerRepository {
	return &PlannerRepository{
		db:  db,
		log: log.With().Str("component", "planner_repository").Logger(),
	}
}

// SequenceRecord represents a sequence in the database.
type SequenceRecord struct {
	ID            int64
	PortfolioHash string
	SequenceData  string // JSON
	PatternType   string
	Depth         int
	Priority      float64
	Completed     bool
	CreatedAt     time.Time
}

// EvaluationRecord represents an evaluation in the database.
type EvaluationRecord struct {
	ID            int64
	SequenceHash  string
	PortfolioHash string
	EndScore      float64
	DeltaScore    float64
	Metrics       string // JSON
	Completed     bool
	CreatedAt     time.Time
}

// BestResultRecord represents the best result in the database.
type BestResultRecord struct {
	ID            int64
	PortfolioHash string
	SequenceHash  string
	PlanData      string // JSON
	Score         float64
	CreatedAt     time.Time
	UpdatedAt     time.Time
}

// InsertSequence inserts a new sequence into the database.
func (r *PlannerRepository) InsertSequence(
	portfolioHash string,
	sequence domain.ActionSequence,
) (int, error) {
	sequenceData, err := json.Marshal(sequence)
	if err != nil {
		return 0, fmt.Errorf("failed to marshal sequence: %w", err)
	}

	result, err := r.db.Exec(`
		INSERT INTO sequences (portfolio_hash, sequence_data, pattern_type, depth, priority, completed, created_at)
		VALUES (?, ?, ?, ?, ?, ?, ?)
	`, portfolioHash, string(sequenceData), sequence.PatternType, sequence.Depth, sequence.Priority, false, time.Now())

	if err != nil {
		return 0, fmt.Errorf("failed to insert sequence: %w", err)
	}

	id, err := result.LastInsertId()
	if err != nil {
		return 0, fmt.Errorf("failed to get last insert id: %w", err)
	}

	r.log.Debug().
		Int64("id", id).
		Str("portfolio_hash", portfolioHash).
		Str("pattern_type", sequence.PatternType).
		Msg("Inserted sequence")

	return int(id), nil
}

// GetSequence retrieves a sequence by ID.
func (r *PlannerRepository) GetSequence(id int64) (*domain.ActionSequence, error) {
	var record SequenceRecord
	err := r.db.QueryRow(`
		SELECT id, portfolio_hash, sequence_data, pattern_type, depth, priority, completed, created_at
		FROM sequences
		WHERE id = ?
	`, id).Scan(
		&record.ID,
		&record.PortfolioHash,
		&record.SequenceData,
		&record.PatternType,
		&record.Depth,
		&record.Priority,
		&record.Completed,
		&record.CreatedAt,
	)

	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("failed to get sequence: %w", err)
	}

	var sequence domain.ActionSequence
	if err := json.Unmarshal([]byte(record.SequenceData), &sequence); err != nil {
		return nil, fmt.Errorf("failed to unmarshal sequence: %w", err)
	}

	return &sequence, nil
}

// ListSequencesByPortfolioHash retrieves all sequences for a portfolio hash.
func (r *PlannerRepository) ListSequencesByPortfolioHash(
	portfolioHash string,
	limit int,
) ([]SequenceRecord, error) {
	query := `
		SELECT id, portfolio_hash, sequence_data, pattern_type, depth, priority, completed, created_at
		FROM sequences
		WHERE portfolio_hash = ?
		ORDER BY priority DESC, created_at DESC
	`
	if limit > 0 {
		query += fmt.Sprintf(" LIMIT %d", limit)
	}

	rows, err := r.db.Query(query, portfolioHash)
	if err != nil {
		return nil, fmt.Errorf("failed to list sequences: %w", err)
	}
	defer rows.Close()

	var records []SequenceRecord
	for rows.Next() {
		var record SequenceRecord
		if err := rows.Scan(
			&record.ID,
			&record.PortfolioHash,
			&record.SequenceData,
			&record.PatternType,
			&record.Depth,
			&record.Priority,
			&record.Completed,
			&record.CreatedAt,
		); err != nil {
			return nil, fmt.Errorf("failed to scan sequence: %w", err)
		}
		records = append(records, record)
	}

	return records, nil
}

// GetPendingSequences retrieves sequences that haven't been evaluated yet.
func (r *PlannerRepository) GetPendingSequences(
	portfolioHash string,
	limit int,
) ([]SequenceRecord, error) {
	query := `
		SELECT id, portfolio_hash, sequence_data, pattern_type, depth, priority, completed, created_at
		FROM sequences
		WHERE portfolio_hash = ? AND completed = 0
		ORDER BY priority DESC, created_at ASC
	`
	if limit > 0 {
		query += fmt.Sprintf(" LIMIT %d", limit)
	}

	rows, err := r.db.Query(query, portfolioHash)
	if err != nil {
		return nil, fmt.Errorf("failed to get pending sequences: %w", err)
	}
	defer rows.Close()

	var records []SequenceRecord
	for rows.Next() {
		var record SequenceRecord
		if err := rows.Scan(
			&record.ID,
			&record.PortfolioHash,
			&record.SequenceData,
			&record.PatternType,
			&record.Depth,
			&record.Priority,
			&record.Completed,
			&record.CreatedAt,
		); err != nil {
			return nil, fmt.Errorf("failed to scan sequence: %w", err)
		}
		records = append(records, record)
	}

	return records, nil
}

// MarkSequenceCompleted marks a sequence as completed.
func (r *PlannerRepository) MarkSequenceCompleted(id int64) error {
	_, err := r.db.Exec(`UPDATE sequences SET completed = 1 WHERE id = ?`, id)
	if err != nil {
		return fmt.Errorf("failed to mark sequence completed: %w", err)
	}

	r.log.Debug().Int64("id", id).Msg("Marked sequence as completed")
	return nil
}

// DeleteSequencesByPortfolioHash deletes all sequences for a portfolio hash.
func (r *PlannerRepository) DeleteSequencesByPortfolioHash(portfolioHash string) error {
	result, err := r.db.Exec(`DELETE FROM sequences WHERE portfolio_hash = ?`, portfolioHash)
	if err != nil {
		return fmt.Errorf("failed to delete sequences: %w", err)
	}

	rowsAffected, _ := result.RowsAffected()
	r.log.Info().
		Str("portfolio_hash", portfolioHash).
		Int64("rows_deleted", rowsAffected).
		Msg("Deleted sequences")

	return nil
}

// InsertEvaluation inserts a new evaluation into the database.
func (r *PlannerRepository) InsertEvaluation(
	evaluation domain.EvaluationResult,
) error {
	metricsData, err := json.Marshal(evaluation.ScoreBreakdown)
	if err != nil {
		return fmt.Errorf("failed to marshal metrics: %w", err)
	}

	// Calculate delta score (improvement over baseline)
	// For now we'll use 0.0 as delta since we don't have baseline score here
	deltaScore := 0.0

	_, err = r.db.Exec(`
		INSERT INTO evaluations (sequence_hash, portfolio_hash, end_score, delta_score, metrics, completed, created_at)
		VALUES (?, ?, ?, ?, ?, ?, ?)
	`, evaluation.SequenceHash, evaluation.PortfolioHash, evaluation.EndScore, deltaScore,
		string(metricsData), true, time.Now())

	if err != nil {
		return fmt.Errorf("failed to insert evaluation: %w", err)
	}

	r.log.Debug().
		Str("sequence_hash", evaluation.SequenceHash).
		Str("portfolio_hash", evaluation.PortfolioHash).
		Float64("end_score", evaluation.EndScore).
		Msg("Inserted evaluation")

	return nil
}

// GetEvaluation retrieves an evaluation by sequence hash.
func (r *PlannerRepository) GetEvaluation(sequenceHash string) (*domain.EvaluationResult, error) {
	var record EvaluationRecord
	err := r.db.QueryRow(`
		SELECT id, sequence_hash, portfolio_hash, end_score, delta_score, metrics, completed, created_at
		FROM evaluations
		WHERE sequence_hash = ?
	`, sequenceHash).Scan(
		&record.ID,
		&record.SequenceHash,
		&record.PortfolioHash,
		&record.EndScore,
		&record.DeltaScore,
		&record.Metrics,
		&record.Completed,
		&record.CreatedAt,
	)

	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("failed to get evaluation: %w", err)
	}

	var scoreBreakdown map[string]float64
	if err := json.Unmarshal([]byte(record.Metrics), &scoreBreakdown); err != nil {
		return nil, fmt.Errorf("failed to unmarshal metrics: %w", err)
	}

	evaluation := &domain.EvaluationResult{
		SequenceHash:   record.SequenceHash,
		EndScore:       record.EndScore,
		ScoreBreakdown: scoreBreakdown,
	}

	return evaluation, nil
}

// ListEvaluationsByPortfolioHash retrieves all evaluations for a portfolio hash.
func (r *PlannerRepository) ListEvaluationsByPortfolioHash(
	portfolioHash string,
) ([]EvaluationRecord, error) {
	rows, err := r.db.Query(`
		SELECT id, sequence_hash, portfolio_hash, end_score, delta_score, metrics, completed, created_at
		FROM evaluations
		WHERE portfolio_hash = ?
		ORDER BY end_score DESC, created_at DESC
	`, portfolioHash)

	if err != nil {
		return nil, fmt.Errorf("failed to list evaluations: %w", err)
	}
	defer rows.Close()

	var records []EvaluationRecord
	for rows.Next() {
		var record EvaluationRecord
		if err := rows.Scan(
			&record.ID,
			&record.SequenceHash,
			&record.PortfolioHash,
			&record.EndScore,
			&record.DeltaScore,
			&record.Metrics,
			&record.Completed,
			&record.CreatedAt,
		); err != nil {
			return nil, fmt.Errorf("failed to scan evaluation: %w", err)
		}
		records = append(records, record)
	}

	return records, nil
}

// DeleteEvaluationsByPortfolioHash deletes all evaluations for a portfolio hash.
func (r *PlannerRepository) DeleteEvaluationsByPortfolioHash(portfolioHash string) error {
	result, err := r.db.Exec(`DELETE FROM evaluations WHERE portfolio_hash = ?`, portfolioHash)
	if err != nil {
		return fmt.Errorf("failed to delete evaluations: %w", err)
	}

	rowsAffected, _ := result.RowsAffected()
	r.log.Info().
		Str("portfolio_hash", portfolioHash).
		Int64("rows_deleted", rowsAffected).
		Msg("Deleted evaluations")

	return nil
}

// UpsertBestResult inserts or updates the best result for a portfolio hash.
func (r *PlannerRepository) UpsertBestResult(
	portfolioHash string,
	result domain.EvaluationResult,
	sequence domain.ActionSequence,
) error {
	// Marshal the sequence (plan data)
	sequenceData, err := json.Marshal(sequence)
	if err != nil {
		return fmt.Errorf("failed to marshal sequence: %w", err)
	}

	// Check if a record exists
	var existingID int64
	err = r.db.QueryRow(`
		SELECT id FROM best_result WHERE portfolio_hash = ?
	`, portfolioHash).Scan(&existingID)

	now := time.Now()

	if err == sql.ErrNoRows {
		// Insert new record
		_, err = r.db.Exec(`
			INSERT INTO best_result (portfolio_hash, sequence_hash, plan_data, score, created_at, updated_at)
			VALUES (?, ?, ?, ?, ?, ?)
		`, portfolioHash, result.SequenceHash, string(sequenceData), result.EndScore, now, now)

		if err != nil {
			return fmt.Errorf("failed to insert best result: %w", err)
		}

		r.log.Info().
			Str("portfolio_hash", portfolioHash).
			Str("sequence_hash", result.SequenceHash).
			Float64("score", result.EndScore).
			Msg("Inserted best result")
	} else if err == nil {
		// Update existing record
		_, err = r.db.Exec(`
			UPDATE best_result
			SET sequence_hash = ?, plan_data = ?, score = ?, updated_at = ?
			WHERE portfolio_hash = ?
		`, result.SequenceHash, string(sequenceData), result.EndScore, now, portfolioHash)

		if err != nil {
			return fmt.Errorf("failed to update best result: %w", err)
		}

		r.log.Info().
			Str("portfolio_hash", portfolioHash).
			Str("sequence_hash", result.SequenceHash).
			Float64("score", result.EndScore).
			Msg("Updated best result")
	} else {
		return fmt.Errorf("failed to check existing best result: %w", err)
	}

	return nil
}

// GetBestResult retrieves the best result for a portfolio hash.
func (r *PlannerRepository) GetBestResult(portfolioHash string) (*domain.HolisticPlan, error) {
	var record BestResultRecord
	err := r.db.QueryRow(`
		SELECT id, portfolio_hash, sequence_hash, plan_data, score, created_at, updated_at
		FROM best_result
		WHERE portfolio_hash = ?
	`, portfolioHash).Scan(
		&record.ID,
		&record.PortfolioHash,
		&record.SequenceHash,
		&record.PlanData,
		&record.Score,
		&record.CreatedAt,
		&record.UpdatedAt,
	)

	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("failed to get best result: %w", err)
	}

	var plan domain.HolisticPlan
	if err := json.Unmarshal([]byte(record.PlanData), &plan); err != nil {
		return nil, fmt.Errorf("failed to unmarshal plan: %w", err)
	}

	return &plan, nil
}

// DeleteBestResult deletes the best result for a portfolio hash.
func (r *PlannerRepository) DeleteBestResult(portfolioHash string) error {
	_, err := r.db.Exec(`DELETE FROM best_result WHERE portfolio_hash = ?`, portfolioHash)
	if err != nil {
		return fmt.Errorf("failed to delete best result: %w", err)
	}

	r.log.Info().
		Str("portfolio_hash", portfolioHash).
		Msg("Deleted best result")

	return nil
}

// CountSequences returns the total number of sequences for a portfolio hash.
func (r *PlannerRepository) CountSequences(portfolioHash string) (int, error) {
	var count int
	err := r.db.QueryRow(`
		SELECT COUNT(*) FROM sequences WHERE portfolio_hash = ?
	`, portfolioHash).Scan(&count)

	if err != nil {
		return 0, fmt.Errorf("failed to count sequences: %w", err)
	}

	return count, nil
}

// CountPendingSequences returns the number of pending sequences for a portfolio hash.
func (r *PlannerRepository) CountPendingSequences(portfolioHash string) (int, error) {
	var count int
	err := r.db.QueryRow(`
		SELECT COUNT(*) FROM sequences WHERE portfolio_hash = ? AND completed = 0
	`, portfolioHash).Scan(&count)

	if err != nil {
		return 0, fmt.Errorf("failed to count pending sequences: %w", err)
	}

	return count, nil
}

// CountEvaluations returns the total number of evaluations for a portfolio hash.
func (r *PlannerRepository) CountEvaluations(portfolioHash string) (int, error) {
	var count int
	err := r.db.QueryRow(`
		SELECT COUNT(*) FROM evaluations WHERE portfolio_hash = ?
	`, portfolioHash).Scan(&count)

	if err != nil {
		return 0, fmt.Errorf("failed to count evaluations: %w", err)
	}

	return count, nil
}
